src/main/java/microsoft/exchange/webservices/data/util/DateTimeParser.java

```diff
@@ -0,0 +1,101 @@
+ **************************************************************************/
+
+package microsoft.exchange.webservices.data.util;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+
+public class DateTimeParser {
+
+  private final SimpleDateFormat[] dateTimeFormats = {
+      new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssX"),
+      new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSX"),
+      new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss"),
+      new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS"),
+      new SimpleDateFormat("yyyy-MM-ddX"),
+      new SimpleDateFormat("yyyy-MM-dd")
+  };
+
+  private final SimpleDateFormat[] dateFormats = {
+      new SimpleDateFormat("yyyy-MM-ddX"),
+      new SimpleDateFormat("yyyy-MM-dd")
+  };
+
+
+  public DateTimeParser() {
+    // Set default timezone of the formats to UTC, which will be used when the date string doesn't supply a
+    // timezone itself.
+
+    for (SimpleDateFormat format : dateTimeFormats) {
+      format.setTimeZone(TimeZone.getTimeZone("UTC"));
+    }
+
+    for (SimpleDateFormat format : dateFormats) {
+      format.setTimeZone(TimeZone.getTimeZone("UTC"));
+    }
+  }
+
+  /**
+   * Converts a date time string to local date time.
+   *
+   * Note: this method also allows dates without times, in which case the time will be 00:00:00 in the
+   * supplied timezone. UTC timezone will be assumed if no timezone is supplied.
+   *
+   * @param value The string value to parse.
+   * @return The parsed {@link Date}.
+   *
+   * @throws java.lang.IllegalArgumentException If string can not be parsed.
+   */
+  public Date convertDateTimeStringToDate(String value) {
+    return parseInternal(value, false);
+  }
+
+  /**
+   * Converts a date string to local date time.
+   *
+   * UTC timezone will be assumed if no timezone is supplied.
+   *
+   * @param value The string value to parse.
+   * @return The parsed {@link Date}.
+   *
+   * @throws java.lang.IllegalArgumentException If string can not be parsed.
+   */
+  public Date convertDateStringToDate(String value) {
+    return parseInternal(value, true);
+  }
+
+  private Date parseInternal(String value, boolean dateOnly) {
+    String originalValue = value;
+
+    if (value == null || value.isEmpty()) {
+      return null;
+    } else {
+      if (value.endsWith("z")) {
+        // This seems to be an edge case. Let's uppercase the Z to be sure.
+        value = value.substring(0, value.length() - 1) + "Z";
+      }
+
+      SimpleDateFormat[] formats = dateOnly ? dateFormats : dateTimeFormats;
+      for (SimpleDateFormat format : formats) {
+        try {
+          return format.parse(value);
+        } catch (ParseException e) {
+          // Ignore and try the next pattern.
+        }
+      }
+    }
+
+    throw new IllegalArgumentException(
```

Comment:
Is there a reason for throwing a different kind of exception here?  IllegalArgumentException vs. ParseException?  I generally prefer to use illegal argument kind of exceptions to reflect programming errors where exceptions based on runtime data would use ParseException (or FormatException in .NET).  Thoughts?